File: backend/src/services/qrService.ts
import QRCode from 'qrcode';
import { v4 as uuidv4 } from 'uuid';

interface QrData {
  orderId: string;
  type: 'seller_pickup' | 'courier_delivery';
  code: string;
  timestamp: number;
}

class QrService {
  async generateSellerQr(orderId: string): Promise<{ qrCode: string; code: string }> {
    const code = uuidv4().substring(0, 8).toUpperCase();
    
    const qrData: QrData = {
      orderId,
      type: 'seller_pickup',
      code,
      timestamp: Date.now(),
    };

    const qrCode = await QRCode.toDataURL(JSON.stringify(qrData), {
      width: 300,
      margin: 2,
      color: {
        dark: '#000000',
        light: '#FFFFFF',
      },
    });

    return { qrCode, code };
  }

  async generateCourierQr(orderId: string): Promise<{ qrCode: string; code: string }> {
    const code = uuidv4().substring(0, 8).toUpperCase();
    
    const qrData: QrData = {
      orderId,
      type: 'courier_delivery',
      code,
      timestamp: Date.now(),
    };

    const qrCode = await QRCode.toDataURL(JSON.stringify(qrData), {
      width: 300,
      margin: 2,
      color: {
        dark: '#000000',
        light: '#FFFFFF',
      },
    });

    return { qrCode, code };
  }

  parseQrCode(qrString: string): QrData | null {
    try {
      const data = JSON.parse(qrString) as QrData;
      
      if (!data.orderId || !data.type || !data.code) {
        return null;
      }

      return data;
    } catch {
      return null;
    }
  }

  validateQrCode(qrData: QrData, expectedOrderId: string, expectedType: 'seller_pickup' | 'courier_delivery'): boolean {
    if (qrData.orderId !== expectedOrderId) {
      return false;
    }

    if (qrData.type !== expectedType) {
      return false;
    }

    const maxAge = 24 * 60 * 60 * 1000;
    if (Date.now() - qrData.timestamp > maxAge) {
      return false;
    }

    return true;
  }
}

export const qrService = new QrService();
export default qrService;

File: frontend/lib/services/qr_service.dart
import 'dart:convert';

class QrData {
  final String orderId;
  final String type;
  final String code;
  final int timestamp;

  QrData({
    required this.orderId,
    required this.type,
    required this.code,
    required this.timestamp,
  });

  factory QrData.fromJson(Map<String, dynamic> json) {
    return QrData(
      orderId: json['orderId'] as String,
      type: json['type'] as String,
      code: json['code'] as String,
      timestamp: json['timestamp'] as int,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'orderId': orderId,
      'type': type,
      'code': code,
      'timestamp': timestamp,
    };
  }

  bool get isExpired {
    final maxAge = 24 * 60 * 60 * 1000;
    return DateTime.now().millisecondsSinceEpoch - timestamp > maxAge;
  }
}

class QrService {
  static final QrService _instance = QrService._internal();
  factory QrService() => _instance;
  QrService._internal();

  QrData? parseQrCode(String qrString) {
    try {
      final json = jsonDecode(qrString) as Map<String, dynamic>;
      return QrData.fromJson(json);
    } catch (e) {
      return null;
    }
  }

  bool validateQrCode(QrData qrData, String expectedOrderId, String expectedType) {
    if (qrData.orderId != expectedOrderId) {
      return false;
    }

    if (qrData.type != expectedType) {
      return false;
    }

    if (qrData.isExpired) {
      return false;
    }

    return true;
  }

  bool isValidPickupQr(String qrString, String orderId) {
    final qrData = parseQrCode(qrString);
    if (qrData == null) return false;
    return validateQrCode(qrData, orderId, 'seller_pickup');
  }

  bool isValidDeliveryQr(String qrString, String orderId) {
    final qrData = parseQrCode(qrString);
    if (qrData == null) return false;
    return validateQrCode(qrData, orderId, 'courier_delivery');
  }
}
